### PR TITLE
A7 : double.Parse

### PR DESCRIPTION
**Pull request by RaaneshKumar**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -3,16 +3,91 @@
...
+      }
+
+      /// <summary>Checks if the user input is valid</summary>
+      /// <param name="s">user input</param>
+      /// <returns>Returns true if the input is valid</returns>
+      public static bool IsValid (string s) {
+         if (s.Any (x => x != '+' && x != '-'                 // Checks if all are valid characters.
+                      && x != '.' && x != 'e'
+                      && !char.IsDigit (x))) return false;
+         if (s.All (x => !char.IsDigit (x))) return false;    // Checks if atleast one digit is present.
```
Comment (priyarathna) [inline Program.cs]: When this condition will be hit? Can you give me an example?

File: Program.cs [hunk 2]
```diff
@@ -3,16 +3,91 @@
...
+         Console.Write ("Enter a value to be converted into double: ");
+         string input = Console.ReadLine ().ToLower ();
+         Console.WriteLine (IsValid (input) ? Parse (input) : "Invalid Input");
+      }
+
+      /// <summary>Checks if the user input is valid</summary>
+      /// <param name="s">user input</param>
+      /// <returns>Returns true if the input is valid</returns>
+      public static bool IsValid (string s) {
+         if (s.Any (x => x != '+' && x != '-'                 // Checks if all are valid characters.
```
Comment (priyarathna) [inline Program.cs]: For better readability, can be simplified as:
var validChars = new char[] { '+', '-', '.', 'e' };
if (s.Any (c => !validChars.Contains (c) && !char.IsDigit (c))) return false;

File: Program.cs [hunk 3]
```diff
@@ -3,16 +3,91 @@
...
+                      && x != '.' && x != 'e'
+                      && !char.IsDigit (x))) return false;
+         if (s.All (x => !char.IsDigit (x))) return false;    // Checks if atleast one digit is present.
+         if (s[0] == 'e' || s[^1] is 'e') return false;       // Checks if first and last characters are valid.
+         if ((s.Where (x => x == '.').Count () > 1) ||        // Checks if multiple decimal points are present.
```
Comment (priyarathna) [inline Program.cs]: Can be simplified as:
if (s.Count (x => x == '.') > 1 ||         // Checks if multiple decimal points are present.
    (s.Count (x => x == 'e') > 1)) return false; // Checks if multiple E's are present.

File: Program.cs [hunk 4]
```diff
@@ -3,16 +3,91 @@
...
+             (s.Where (x => x == 'e').Count () > 1)) return false; // Checks if multiple E's are present.
+         if (!ExpressionCheck ('+') || !ExpressionCheck ('-')) return false;
+         return true;
+
+         /// <summary>Checks if the '+' or '-' in user input is unary</summary>
+         /// <param name="a">char '+' or '-'</param>
+         bool ExpressionCheck (char a) {
+            switch (s.Where (x => x == a).Count ()) {
```
Comment (priyarathna) [inline Program.cs]: Can be simplified as:
switch (s.Count (x => x == a))

File: Program.cs [hunk 5]
```diff
@@ -3,16 +3,91 @@
...
+      /// <summary>Checks if the input has 'e' in it</summary>
+      /// <returns></returns>
+      static bool ContainsE (string s) => s.Any (x => x == 'e');
+
+      /// <summary>Converts the string to double if E is present</summary>
+      /// <returns>Returns the converted double value</returns>
+      static double GetSolvedE (string s) {
+         double firstPart, secPart;
+         string[] parts = s.Split ('e');
+
```
Comment (priyarathna) [inline Program.cs]: No need of 'empty line' here. Apply it wherever necessary.

File: Program.cs [hunk 6]
```diff
@@ -3,16 +3,91 @@
...
+         firstPart = ContainsPoint (parts[0]) ? GetSolvedPoint (parts[0]) : int.Parse (parts[0]);
+         secPart = ContainsPoint (parts[1]) ? GetSolvedPoint (parts[1]) : int.Parse (parts[1]);
+         return firstPart * Math.Pow (10, secPart);
+      }
+
+      /// <summary>Checks if the input has a decimal point in it</summary>
+      /// <returns>Returns true if decimal point is present</returns>
+      static bool ContainsPoint (string s) => s.Any (x => x == '.');
+
+      /// <summary>Converts the string to double if decimal pint is present.</summary>
```
Comment (priyarathna) [inline Program.cs]: point spell wrong.

File: Program.cs [hunk 7]
```diff
@@ -3,16 +3,91 @@
...
+      /// <summary>Coverts the string to a double based on the type of input</summary>
+      /// <returns>Returns the converted double value</returns>
+      public static double Parse (string s) {
+         if (ContainsE (s)) return GetSolvedE (s);
+         if (!ContainsPoint (s)) return int.Parse (s); // Means an integer
+         return GetSolvedPoint (s);
+      }
+
+      /// <summary>Checks if the input has 'e' in it</summary>
+      /// <returns></returns>
```
Comment (priyarathna) [inline Program.cs]: Include comment for 'return'.

**Review by priyarathna — CHANGES_REQUESTED** [review]

File: Program.cs (hunk 1 above)
Comment (RaaneshKumar) [reply Program.cs]: Was written for inputs like "+e." where all characters are valid but none of them is a digit. 
But removed this condition and added extra invalid conditions such that it also covers inputs including these and other necessary ones.

**Review by priyarathna — APPROVED** [review]

**Pull request closed without merging**